File: kotlin/src/test/kotlin/tracking/processing/PreprocessorTest.kt
package tracking.processing

import org.junit.jupiter.api.DynamicTest.dynamicTest
import org.junit.jupiter.api.TestFactory
import testUtils.getCreatorData
import testUtils.getUrl
import tracking.contents.ProcessedItem
import tracking.website.StandardStrategy
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

class PreprocessorTest {
    private lateinit var subject: Preprocessor

    @BeforeTest
    fun beforeTest() {
        subject = Preprocessor()
    }

    @TestFactory
    fun `Cleaner regexes are working`() = mapOf(
        "***open***" to "open",
        "!closed!" to "closed",
        " ❗&nbsp;" to " ! ", // Unicode NBSP, emoticon !, HTML entity NBSP
        "\t" to " ",
    ).map { (input, expected) ->
        dynamicTest("Test input: '${input}'") {
            val testItem = ProcessedItem(getCreatorData(), getUrl(), StandardStrategy, input)

            subject.preprocess(testItem)

            assertEquals(expected, testItem.contents)
        }
    }

    @Test
    fun `Input gets converted to lowercase`() {
        val testItem = ProcessedItem(getCreatorData(), getUrl(), StandardStrategy, "AaBbCcDdEeFf")

        subject.preprocess(testItem)

        assertEquals("aabbccddeeff", testItem.contents)
    }

    @TestFactory
    fun `Creator aliases are getting replaced with the name placeholder`() = listOf(
        Triple(
            "An Intergalactic House of Pancakes work",
            listOf("Intergalactic House of Pancakes"),
            "an CREATOR_NAME work",
        ),
        Triple(
            "An Intergalactic House of Pancake's work",
            listOf("Intergalactic House of Pancakes"),
            "an CREATOR_NAME work",
        ),
        Triple(
            "About Intergalactic Pancake's work",
            listOf("Intergalactic Pancake"),
            "about CREATOR_NAME's work",
        ),
        Triple(
            // Multiple aliases, 's form, case-insensitive, "creator" in aliases
            "asdf Studio's uiop Creator asdf Studios zxcv",
            listOf("StUdIoS", "cReatOR"),
            "asdf CREATOR_NAME uiop CREATOR_NAME asdf CREATOR_NAME zxcv",
        ),
    ).map { (input, aliases, expected) ->
        dynamicTest("Test input: '${input}'") {
            val testItem = ProcessedItem(getCreatorData(aliases = aliases), getUrl(), StandardStrategy, input)

            subject.preprocess(testItem)

            assertEquals(expected, testItem.contents)
        }
    }

    @TestFactory
    fun `False positives are being removed`() = mapOf(
        "even though you're closed for commissions" to "",
        "while mine commissions are open" to "",
        "if my quotes open" to "",
        "- art commissions are open" to "- are open",
        "after the commissions close" to "",
        "although comms are closed" to "",
        "as soon as we're open" to "",
        "next commissions opening" to "",
        "commissions: open January" to "",
        "The Creator is now opening for quotes a few weeks before commission slots open" to "",
        "when do you open for" to "",
        "when i'm taking" to "",
        "when will you start taking new commissions?" to "?",
    ).map { (input, expected) ->
        dynamicTest("Test input: '${input}'") {
            val testItem = ProcessedItem(getCreatorData(aliases = listOf("The Creator")), getUrl(), StandardStrategy, input)

            subject.preprocess(testItem)

            assertEquals(expected, testItem.contents)
        }
    }
}

File: symfony/tests/Tracking/PreprocessorTest.php
<?php

declare(strict_types=1);

namespace App\Tests\Tracking;

use App\Tests\TestUtils\Cases\FuzzrakeTestCase;
use App\Tests\TestUtils\DataDefinitions;
use App\Tracking\Patterns\Patterns;
use App\Tracking\Patterns\RegexesLoader;
use App\Tracking\Preprocessor;
use Override;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\Attributes\Small;
use Veelkoov\Debris\StringList;

#[Small]
class PreprocessorTest extends FuzzrakeTestCase
{
    private Preprocessor $subject;

    #[Override]
    public function setUp(): void
    {
        /* @phpstan-ignore argument.type */
        $regexesLoader = new RegexesLoader(DataDefinitions::get('tracking.yaml', 'tracking'));
        $this->subject = new Preprocessor(new Patterns($regexesLoader));
    }

    /**
     * @return list<array{string, string}>
     */
    public static function cleanerRegexesAreWorkingDataProvider(): array
    {
        return [
            ['***open***', 'open'],
            ['!closed!', 'closed'],
            [' ❗&nbsp;', ' ! '], // Unicode NBSP, emoticon !, HTML entity NBSP
            ["\t", ' '],
        ];
    }

    #[DataProvider('cleanerRegexesAreWorkingDataProvider')]
    public function testCleanerRegexesAreWorking(string $input, string $expected): void
    {
        $result = $this->subject->preprocess($input, new StringList());

        self::assertSame($expected, $result);
    }

    public function testInputGetsConvertedToLowercase(): void
    {
        $result = $this->subject->preprocess('AaBbCcDdEeFf', new StringList());

        self::assertSame('aabbccddeeff', $result);
    }

    /**
     * @return list<array{string, StringList, string}>
     */
    public static function creatorAliasesAreGettingReplacedWithTheNamePlaceholderDataProvider(): array
    {
        return [
            [
                'An Intergalactic House of Pancakes work',
                new StringList(['Intergalactic House of Pancakes']),
                'an CREATOR_NAME work',
            ],
            [
                "An Intergalactic House of Pancake's work",
                new StringList(['Intergalactic House of Pancakes']),
                'an CREATOR_NAME work',
            ],
            [
                "About Intergalactic Pancake's work",
                new StringList(['Intergalactic Pancake']),
                "about CREATOR_NAME's work",
            ],
            [
                // Multiple aliases, 's form, case-insensitive, "creator" in aliases
                "asdf Studio's uiop Creator asdf Studios zxcv",
                new StringList(['StUdIoS', 'cReatOR']),
                'asdf CREATOR_NAME uiop CREATOR_NAME asdf CREATOR_NAME zxcv',
            ],
        ];
    }

    #[DataProvider('creatorAliasesAreGettingReplacedWithTheNamePlaceholderDataProvider')]
    public function testAliasesGetReplacedWithPlaceholder(string $input, StringList $aliases, string $expected): void
    {
        $result = $this->subject->preprocess($input, new StringList($aliases));

        self::assertSame($expected, $result);
    }

    /**
     * @return list<array{string, string}>
     */
    public static function falsePositivesAreBeingRemovedDataProvider(): array
    {
        return [
            ["even though you're closed for commissions", 'FALSE_POSITIVE'],
            ['while mine commissions are open', 'FALSE_POSITIVE'],
            ['if my quotes open', 'FALSE_POSITIVE'],
            ['- art commissions are open', '- FALSE_POSITIVE are open'],
            ['after the commissions close', 'FALSE_POSITIVE'],
            ['although comms are closed', 'FALSE_POSITIVE'],
            ["as soon as we're open", 'FALSE_POSITIVE'],
            ['next commissions opening', 'FALSE_POSITIVE'],
            ['commissions: open January', 'FALSE_POSITIVE'],
            ['The Creator is now opening for quotes a few weeks before commission slots open', 'FALSE_POSITIVE'],
            ['when do you open for', 'FALSE_POSITIVE'],
            ["when i'm taking", 'FALSE_POSITIVE'],
            ['when will you start taking new commissions?', 'FALSE_POSITIVE?'],
        ];
    }

    #[DataProvider('falsePositivesAreBeingRemovedDataProvider')]
    public function testFalsePositivesAreBeingRemoved(string $input, string $expected): void
    {
        $result = $this->subject->preprocess($input, StringList::of('The Creator'));

        self::assertSame($expected, $result);
    }
}
